src/gc-heap-snapshot.cpp

```diff
@@ -0,0 +1,420 @@
+
+const int k_node_number_of_fields = 7;
+struct Node {
+    size_t type; // TODO: point at actual type here?
+    string name;
+    size_t id; // (vilterp) the memory address, right?
+    size_t self_size;
+    size_t trace_node_id;  // This is ALWAYS 0 in Javascript heap-snapshots.
+    // whether the from_node is attached or dettached from the main application state
+    // TODO: .... meaning not yet understood.
+    // https://github.com/nodejs/from_node/blob/5fd7a72e1c4fbaf37d3723c4c81dce35c149dc84/deps/v8/include/v8-profiler.h#L739-L745
+    int detachedness;  // 0 - unknown,  1 - attached;  2 - detached
+
+    // Book-keeping fields (not used for serialization)
+    vector<Edge> edges; // For asserting that we built the edges in the right order
+};
+
+
+struct StringTable {
+    typedef unordered_map<string, size_t> MapType;
+
+    MapType map;
+    vector<string> strings;
+
+    StringTable() {}
+    StringTable(std::initializer_list<string> strs) : strings(strs) {
+        for (const auto& str : strs) {
+            map.insert({str, map.size()});
+        }
+    }
+
+    size_t find_or_create_string_id(string key) {
+        auto val = map.find(key);
+        if (val == map.end()) {
+            val = map.insert(val, {key, map.size()});
+            strings.push_back(key);
+        }
+        return val->second;
+    }
+
+    void print_json_array(JL_STREAM *stream, bool newlines) {
+        jl_printf(stream, "[");
+        bool first = true;
+        for (const auto &str : strings) {
+            if (first) {
+                first = false;
+            } else {
+                jl_printf(stream, ",");
+                if (newlines) {
+                    jl_printf(stream, "\n");
+                }
+            }
+            // Escape strings for JSON
+            // TODO
+            print_str_escape_json(stream, str);
+        }
+        jl_printf(stream, "]");
+    }
+};
+
+struct HeapSnapshot {
+public:
+
+// private:
+    vector<Node> nodes;
+    // edges are stored on each from_node
+
+    StringTable names;
+    StringTable node_types;
+    StringTable edge_types;
+    unordered_map<void*, size_t> node_ptr_to_index_map;
+
+    size_t num_edges = 0; // For metadata, updated as you add each edge. Needed because edges owned by nodes.
+};
+
+
+// TODO: Do we need to refer to nodes by their index in the from_node array?
+//size_t find_or_create_node_id(HeapSnapshot& snapshot, string key) {
+//    return find_or_insert_iter(snapshot.nodes_map, key)->second;
+//}
+
+HeapSnapshot *g_snapshot = nullptr;
+
+JL_DLLEXPORT int count_nodes = 0;
+JL_DLLEXPORT int count_edges = 0;
+
+JL_DLLEXPORT void jl_gc_take_heap_snapshot(JL_STREAM *stream) {
+    // Enable snapshotting
+    HeapSnapshot snapshot;
+    g_snapshot = &snapshot;
+    gc_heap_snapshot_enabled = true;
+
+    _add_internal_root(&snapshot);
+
+    // Do GC, which will callback into record_edge_to_gc_snapshot()...
+    jl_gc_collect(JL_GC_FULL);
```

Comment:
😱 oh wow, this is a huge find! Thanks @vtjnash! 🎉 Very interesting.

Yeah, we should definitely look into this more.

It would be great if there was a surefire way to trigger a full mark phase, to be sure that this doesn't bitrot as other threading-related changes are made. Maybe we can introduce a helper function for that as part of this PR?

Reply:
Hrmm i think we've done this now, but it still doesn't exactly seem to be working.. Now i'm getting only 6MB snapshots, whereas before they were like 80MB.

If I change the second call to still do `JL_GC_FULL`, then i see the following behavior:
- the first snapshot is still 5MB and runs quickly
- the second and third are both 80MB, and take a little while.

(Apple reports a new julia process as using 113MB of RAM.)

Do you think this still isn't quite right, somehow?